File: apps/web/lib/actions/files.ts
/// 文件库操作(前端 HTTP 客户端)。
///
/// 设计意图:
///   原为 Server Actions,现调用后端 REST API。**契约与原 Server Action 一致**:
///   返回 `{ success, error?, files?, nextCursor?, hasMore? }`,调用方
///   (library-panel)按此判定,无需改动。
///   后端端点在阶段 4/任务 20 实现;未就绪时走 catch 返回 `{ success: false }`。

import { apiFetch } from '@/lib/api-client'
import type { LibraryFile } from '@/lib/db/schema'

/// 与原项目 FilesListCursor 对齐:按 (updatedAt, id) 排序的分页游标
export type FilesListCursor = {
  updatedAt: string
  id: string
}

/// 与原项目对齐:LibraryFile + 签名后的可访问 url
export type LibraryFileItem = LibraryFile & {
  key: string
  url: string
}

export async function listFiles({
  limit = 20,
  cursor
}: {
  limit?: number
  cursor?: FilesListCursor | null
} = {}): Promise<{
  success: boolean
  files?: LibraryFileItem[]
  nextCursor?: FilesListCursor | null
  hasMore?: boolean
  error?: string
}> {
  try {
    const params = new URLSearchParams({ limit: String(limit) })
    if (cursor) params.set('cursor', JSON.stringify(cursor))
    return await apiFetch(`/api/files?${params}`)
  } catch (error) {
    return {
      success: false,
      files: [],
      error: error instanceof Error ? error.message : 'Failed to load files'
    }
  }
}

export async function searchFiles({
  query,
  limit = 20
}: {
  query: string
  limit?: number
}): Promise<{
  success: boolean
  files?: LibraryFileItem[]
  nextCursor?: FilesListCursor | null
  hasMore?: boolean
  error?: string
}> {
  try {
    const params = new URLSearchParams({
      query: encodeURIComponent(query),
      limit: String(limit)
    })
    return await apiFetch(`/api/files?${params}`)
  } catch (error) {
    return {
      success: false,
      files: [],
      error: error instanceof Error ? error.message : 'Failed to search files'
    }
  }
}

export async function deleteFile(
  fileId: string
): Promise<{ success: boolean; error?: string }> {
  try {
    await apiFetch(`/api/files/${fileId}`, { method: 'DELETE' })
    return { success: true }
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : 'Failed to delete file'
    }
  }
}

File: apps/server/src/ryoshi/api/files.py
"""文件库路由:GET /api/files(列表/搜索)、DELETE /api/files/:id。

设计意图:
    对应原项目 lib/actions/files.ts。用户在 Library 面板浏览/搜索/删除
    已上传的文件。文件元信息存在 files 表,实际内容在对象存储(S3/R2)。

    认证: ENABLE_AUTH=false 时共享匿名用户;ENABLE_AUTH=true 时从 JWT 解析。
    分页用游标(updatedAt, id),与笔记路由保持一致。
"""

import json
from datetime import datetime
from typing import Any

from fastapi import APIRouter, Depends, Header, HTTPException, Query
from sqlalchemy import delete as sa_delete
from sqlalchemy import or_, select
from sqlalchemy.ext.asyncio import AsyncSession

from ryoshi.auth import AuthError, AuthUser, resolve_user
from ryoshi.db.engine import get_session
from ryoshi.db.models import LibraryFile

router = APIRouter(prefix="/api/files", tags=["files"])


async def _current_user(authorization: str | None = Header(None)) -> AuthUser:
    try:
        return await resolve_user(authorization, allow_anonymous_fallback=True)
    except AuthError as exc:
        raise HTTPException(status_code=401, detail=str(exc)) from exc


def _file_to_dict(f: LibraryFile) -> dict[str, Any]:
    return {
        "id": f.id,
        "filename": f.filename,
        "objectKey": f.object_key,
        "mediaType": f.media_type,
        "size": f.size,
        "chatId": f.chat_id,
        "createdAt": f.created_at.isoformat() if f.created_at else None,
        "updatedAt": f.updated_at.isoformat() if f.updated_at else None,
        # LibraryFileItem 需要的 key/url 字段
        "key": f.object_key,
    }


@router.get("")
async def list_files(
    limit: int = Query(20, ge=1, le=100),
    cursor: str | None = Query(None),
    query: str | None = Query(None),
    session: AsyncSession = Depends(get_session),
    user: AuthUser = Depends(_current_user),
):
    """分页/搜索文件库。对应原项目 listFiles / searchFiles。"""
    stmt = select(LibraryFile).where(LibraryFile.user_id == user.id)

    if query:
        stmt = stmt.where(
            or_(
                LibraryFile.filename.ilike(f"%{query}%"),
                LibraryFile.media_type.ilike(f"%{query}%"),
            )
        )

    if cursor:
        try:
            c = json.loads(cursor)
            cursor_time = datetime.fromisoformat(c["updatedAt"].replace("Z", "+00:00"))
            cursor_id = c["id"]
            stmt = stmt.where(
                (LibraryFile.updated_at < cursor_time)
                | ((LibraryFile.updated_at == cursor_time) & (LibraryFile.id < cursor_id))
            )
        except (json.JSONDecodeError, KeyError, ValueError):
            raise HTTPException(status_code=400, detail="Invalid cursor") from None

    stmt = stmt.order_by(LibraryFile.updated_at.desc(), LibraryFile.id.desc()).limit(limit + 1)
    result = await session.execute(stmt)
    files = list(result.scalars().all())

    has_more = len(files) > limit
    files = files[:limit]

    next_cursor = None
    if has_more and files:
        last = files[-1]
        next_cursor = {
            "updatedAt": last.updated_at.isoformat() if last.updated_at else "",
            "id": last.id,
        }

    return {
        "success": True,
        "files": [_file_to_dict(f) for f in files],
        "nextCursor": next_cursor,
        "hasMore": has_more,
    }


@router.delete("/{file_id}", status_code=204)
async def delete_file(
    file_id: str,
    session: AsyncSession = Depends(get_session),
    user: AuthUser = Depends(_current_user),
):
    """删除一条文件记录。仅 owner 可操作。注意:只删 DB 记录,不删对象存储里的文件。"""
    file = await session.get(LibraryFile, file_id)
    if file is None or file.user_id != user.id:
        raise HTTPException(status_code=404, detail="File not found")
    await session.execute(sa_delete(LibraryFile).where(LibraryFile.id == file_id))
    await session.commit()
